prime-router/src/main/kotlin/azure/ReportFunction.kt

```diff
@@ -418,4 +448,45 @@ class ReportFunction {
         }
         return outStream.toString()
     }
+
+    /**
+     * Creates a list of [ItemJurisdictionRouting] instances with the report index
+     * and a jurisdiction keyed map with a list of the receiver organizations where
+     * the report was routed.
+     * @param validatedRequest the instance generated while processing the report
+     * @param actionHistory the instance generated while processing the report
+     * @return the report routing for each item broken down by jurisdiction
+     */
+    private fun createItemJurisdictionRouting(
+        validatedRequest: ValidatedRequest,
+        actionHistory: ActionHistory? = null,
+    ): List<ItemJurisdictionRouting> {
+        val routing = mutableListOf<ItemJurisdictionRouting>()
+        validatedRequest.report?. let { report ->
+            // the report has all the submitted items
+            report.itemIndices.forEach { reportIndex ->
+                val trackingId = report.schema.trackingElement?. let {
+                    report.getString(reportIndex, report.schema.trackingElement)
+                }
+                val destinations = mutableMapOf<Organization.Jurisdiction, MutableList<String>>()
+                // find all outgoing reports based on the index and generate a mapped
+                // list of the receiving organizations keyed by the jurisdiction
+                actionHistory?. let { ah ->
+                    ah.itemLineages.filter { il ->
+                        il.parentIndex == reportIndex
```

Comment:
Tom, as written is O(n squared) - you are iterating through the items in the report, and _for each one_ you are doing a filter of all the data in the itemLineages list, which also iterates thru a list!   This will be very slow for large reports.

At the same time, I think you can go ahead and remove the top level divvying by Jurisdiction type (eg FEDERAL, STATE, etc) if you like -- I think that would make it simpler to do this.

So, to do this in O(n) time, I think you can just walk the itemLineages list (only).    List<ItemJurisdictionRouting> should instead be a `Map<Integer, ItemLineageDetails>` where the Integer is the reportIndex, and ItemLineageDetails contains 
- the reportIndex
- the trackingId
- the jurisdiction
- and a destinations: List<String> where each String in destinations is a receiver fullName.
As you iterate thru itemLineages, you either create a new ItemLineageDetails if it doesn't exist, or if it already exists, you add a new receiver to the destinations list.

Be sure to sort the Map by the Integer key before printing the json.

A good test is this, where every other records go to both Pima County and AZ State:

`curl -X POST -H "client:simple_report"  -H "content-type:text/csv" --data-binary "@./src/test/csv_test_files/input/simplereport.csv"  "http://localhost:7071/api/reports"`

Reply:
This is a good catch. I had thought about writing a lookup map with the original PR. With the revelation that you can use report index vs trackingId I see how this is simplified now.